Documentation/ExtensionArchitecture/BestPractises/SoftwareDesignPrinciples.rst

```diff
@@ -0,0 +1,34 @@
+.. include:: /Includes.rst.txt
+.. index:: Extension development; Software Design Principles
+.. _extension-software-design-principles:
+
+==========================
+Software Design Principles
+==========================
+
+The following principles are important to know when you plan to develop
+extensions for TYPO3.
+
+*   `Object-oriented programming (OOP) <https://phptherightway.com/#programming_paradigms>`__:
+    describes how to encapsulate associated real-world aspects to abstract
+    objects in a software.
+*   `Domain-driven design (DDD) <https://leanpub.com/ddd-in-php>`__: The goal of t
+    his approach is to transcribe terms, rules, and actions of the problem
+    at hand in an adequate way.
+*   `Model-view-controller (MVC) <https://code.tutsplus.com/tutorials/mvc-for-noobs--net-10488>`__:
+    This programming paradigm leads to clear isolation of data, control of
+    actions, and logic of interaction.
+*   `Test-driven development (TDD) <https://phptherightway.com/#test_driven_development>`__:
+    This approach is a basic technique for generating code that is stable,
+    resilient to errors and legible - and therefore maintainable.
+*   `Dependency Injection <https://phptherightway.com/#dependency_injection>`__:
+    A software design pattern that allows the removal of hard-coded dependencies
+    and makes it possible to change them.
+*   `Single-responsibility principle (SRP) <https://en.wikipedia.org/wiki/Single-responsibility_principle>`__:
+    computer programming principle that states that "A module should be
+    responsible to one, and only one, actor." The term actor refers to a group
+    (consisting of one or more stakeholders or users) that requires a change
+    in the module.
```

Comment:
```suggestion
The following principles are considered best practices and are good to know when you develop
extensions for TYPO3.

*   `Object-Oriented Programming (OOP) <https://phptherightway.com/#programming_paradigms>`__ organizes software design around data, or objects, rather than the functions and logic of procedural programming.
*   `Domain-Driven Design (DDD) <https://leanpub.com/ddd-in-php>`__ is a collection of principles and patterns that help developers design elegant code around clearly defined conceptual models.
*   `Model-View-Controller (MVC) <https://code.tutsplus.com/tutorials/mvc-for-noobs--net-10488>`__ is a programming paradigm that leads to clear isolation of data, presentation layer, and business logic.
*   `Test-Driven Development (TDD) <https://phptherightway.com/#test_driven_development>`__  is a basic technique for generating code that is stable,
    resilient to errors, and legible — and therefore maintainable.
*   `Dependency Injection <https://phptherightway.com/#dependency_injection>`__ is a software design pattern that adds flexibility by removing the need for hard-coded dependencies.
*   `Separation of Concerns<https://en.wikipedia.org/wiki/Separation_of_concerns>`__ and `Single-responsibility principle (SRP) <https://en.wikipedia.org/wiki/Single-responsibility_principle>`__ are computer programming principles that help developers create modular, coherent, and maintainable software.
```